### ISSUE TITLE
Develop example of switch design pattern

### ISSUE THREAD

**Issue by mcking65** [w3c/aria-practices#243]
This issue is dependent on issue #162.

#### Preview link for work in progress

[Switch example in switch branch](https://rawgit.com/w3c/aria-practices/switch/examples/switch/switch.html)

**Comment by MichielBijl** [comment]
Work on this is done in the “switch” branch. You can follow progress on the [live example](https://rawgit.com/w3c/aria-practices/switch/examples/switch/switch.html).

**Comment by MichielBijl** [comment]
Got [approval to use Carri Craver’s design](https://twitter.com/ccgirl/status/888024002588966912).

**Comment by Owlbertz** [comment]
Because I recently stumbled upon accessibility of a switch, I found this one. I just noticed that the switch can only be operated using space and not enter. What's the reason for this behavior?

**Comment by MichielBijl** [comment]
@Owlbertz because that’s the way native switches and checkboxes behave.

**Comment by annabbott** [comment]
No issues noted.

**Comment by jnurthen** [comment]
Needs to have a greater contrast in the off state to meet WCAG 2.1

**Comment by stes-acc** [comment]
Examples are insufficient.
What if Switch itself has a label inside that dynamically changes based on switch state? (e.g. On/Off)

**Comment by jnurthen** [comment]
@stes-acc Can you submit a separate issue for this. It would be helpful if you could provide a link to an example which demonstrates the feature so there is clarity as to what is being requested.

**Comment by stes-acc** [comment]
Well, an example can be found here [https://openui5.hana.ondemand.com/#/sample/sap.m.sample.Switch/preview](url)

**Comment by jnurthen** [comment]
@stes-acc New issue please.

**Comment by stes-acc** [comment]
>>> All the examples you cited are problematic and violate this guidance for toggles.

We have **never seen** such a _guidance_ written and recommended as official statement from the ARIA group. Can you please point us to the written source of this information?
Nevertheless we will elaborate on your proposals, many thanks.

We don't want or need role=application on body but we need it to correct inappropriate and insufficient Jaws ARIA role recognition and Forms mode treatment for ARIA mainly in IE but also other browsers which is a commonly known Jaws issue. We know that it confuses potentially users but that is the situation. Not we but Jaws has to improve here making those workarounds obsolete.

**Comment by carmacleod** [comment]
@mcking65 
`aria-checked` is a required attribute for `switch`, `checkbox`, etc. We currently have it as an author error, with a fallback value of `false`, in the [Required Properties with Default Values table](https://github.com/w3c/aria/wiki/Required-Properties-with-Default-Values) that we created for https://github.com/w3c/aria/issues/787.

Maybe we should discuss changing `aria-checked` to a supported attribute instead?
@jnurthen a topic for the call?

**Comment by jnurthen** [comment]
> > @jnurthen wrote:
> > Needs to have a greater contrast in the off state to meet WCAG 2.1
> 
> Have changed the design. The on state now passes with 3:1. But the off state still fails with 1.6:1. I’ll have another look at the design.

background: #949494 
would fix it - the question is, is it too dark that someone might mistake it for being "on". My gut feeling is that it is ok but I have no research to back this up :)
If simply darkening the grey is not acceptable then potentially simply having a slightly darker circle or border for the switch and retain the grey interior colour could work as that would still have a 3:1 contrast with the white.

**Comment by MichielBijl** [comment]
That could work 👍 

I was looking at something like this [design on Dribbble](https://dribbble.com/shots/3825183-On-Off-Toggle-Interaction) which shows the state in text too—similar to how Apple does this in iOS.

**Comment by stes-acc** [comment]
Nevertheless there is the need to develop a switch pattern that changes and signals its internal value **if it is NOT a boolean value.**

There are different approaches to accomplish that:

1) Not using the switch role at all but the **slider** role instead with 2 values and use aria-valuetext attribute for the non-boolean value

2) Using the switch role pointing to an **additional live region** that dynamically changes with respect to the switch state, for example:

```
<span id="sl1">Please confirm:</span>
<span id="s1" role="switch" aria-labelledby="sl1" aria-describedby="lr" aria-checked="false" onclick="toggle(event)" onkeyup="checkKey(event)" class="switch red" tabindex="0">
<span id="lr" class="stext" aria-live="polite">Reject</span>
</span>

```
Please comment what to do in such cases.

**Comment by ZoeBijl** [comment]
@stes-acc do you have a working code example we could look at? I’ll have to read the rest of the thread but I’m not quite sure what you mean and whether it’s a switch pattern.

**Comment by stes-acc** [comment]
> If this is about the Open UI switch example you posted earlier; all I see there is that the state is clarified by the text inside the switch. This is similar to how iOS adds an I and O indication within its accessibility features.

As a matter of fact, I and O are NOT separate non-boolean value text but an additional visualization of boolean values for people not knowing which switch position means on and which off.

Regarding th example:

Here is a rough sketch [https://codepen.io/anon/pen/NmqpBW](url)

It does NOT contain a switch-like layout or any sliding animation. It should only demonstrate the principle of dynamically changing the text on a switch when activated. And don't tell me now use a toggle button for that because it looks alike. The requirement **is a switch layout** as in the [Open UI switch example](https://openui5.hana.ondemand.com/#/sample/sap.m.sample.Switch/preview).

**Comment by mcking65** [comment]
Complete in #1892  with commit 594fe37.

@stes-acc I think we have addressed your concern about people not being able to understand which graphical position means "on" and which means "off". Please have a look at the first [Switch Example](https://w3c.github.io/aria-practices/examples/switch/switch.html). BTW, 2 more examples, one made with HTML checkbox and one ith HTML button are in the works and will land soon.
